### ISSUE TITLE
[마일스톤4-Day1] ML 학습용 데이터 수집 강화

### ISSUE THREAD

**Issue by La-vela** [FerryLa/News-Leafletter#56]
[마일스톤4-Day1] ML 학습용 데이터 수집 강화

## 📌 간단 개요
기사 메타데이터 보강 및 ML 학습을 위한 데이터 수집 파이프라인 강화. 키워드 자동 추출 기능 추가 및 데이터 품질 향상.

## 🎯 목표
1. **데이터베이스 스키마 확장**: ML 학습에 필요한 메타데이터 컬럼 추가
2. **키워드 자동 추출**: TF-IDF 기반 핵심 키워드 자동 추출 시스템 구축
3. **데이터 품질 향상**: 메타데이터 누락 최소화 및 수집 통계 모니터링
4. **학습 데이터 확보**: 최소 1,000개 이상의 양질의 기사 데이터 확보

### 1. 스키마 확장
**기존 문제**:
- ML 학습에 필요한 메타데이터 부족
- 기사 품질 측정 지표 없음
- 키워드 정보 수동 관리

**개선 방향**:
- [ ] `articles` 테이블에 신규 컬럼 추가
  ```sql
  ALTER TABLE articles ADD COLUMN content_length INTEGER;
  ALTER TABLE articles ADD COLUMN keyword_extracted TEXT;  -- JSON 형태
  ALTER TABLE articles ADD COLUMN sentiment_score REAL;    -- 선택 사항
  ```
- [ ] 기존 데이터 마이그레이션 (content_length 자동 계산)

### 2. 수집 파이프라인 강화
**기존 문제**:
- RSS 수집 시 전문(content) 누락 빈번
- author, published_at 등 메타데이터 수집 불완전

**개선 방향**:
- [ ] RSS 파서 개선 (content 추출 로직 강화)
- [ ] 메타데이터 필수 필드 검증 로직 추가
- [ ] 수집 실패 시 재시도 로직 구현
- [ ] 최소 1,000개 기사 목표 달성 모니터링

### 3. 키워드 자동 추출
**기존 문제**:
- 키워드 정보가 유저 등록 키워드에만 의존
- 기사별 핵심 키워드 자동 추출 없음

**개선 방향**:
- [ ] TF-IDF 기반 키워드 추출 모듈 구현
  - 제목 + 요약에서 핵심 단어 추출
  - 상위 5-10개 키워드 선정
  - 한국어 불용어(stopwords) 제거
- [ ] JSON 형태로 저장
  ```json
  {
    "keywords": ["비트코인", "ETF", "승인", "SEC", "투자"],
    "scores": [0.85, 0.72, 0.68, 0.55, 0.42]
  }
  ```

### 4. 검증 및 모니터링
**기존 문제**:
- 데이터 수집 현황 파악 어려움
- 품질 체크 수동으로 진행

**개선 방향**:
- [ ] 데이터 수집 통계 자동 생성 스크립트
- [ ] 누락 필드 비율 자동 체크
- [ ] 일일 수집량 모니터링

### 1. ML 학습 준비 완료
- **효과**: 임베딩, 추천, 클러스터링 등 다양한 ML 실험 가능
- **지표**: 1,000+ 기사, 메타데이터 완성도 90% 이상

### 2. 데이터 품질 향상
- **효과**: content_length 측정으로 짧은 기사 자동 필터링
- **지표**: content 누락률 30% → 10% 이하로 개선

### 3. 키워드 기반 분석 가능
- **효과**: 기사별 핵심 키워드 자동 추출로 트렌드 분석 가능
- **지표**: 전체 기사의 90% 이상 키워드 추출 완료

### 4. 자동화된 모니터링
- **효과**: 데이터 수집 현황 실시간 파악 및 문제 조기 발견
- **지표**: 통계 스크립트 실행 시간 < 5초

## 📦 산출물
- [x] **마이그레이션 스크립트**: `scripts/migrate_ml_columns.py`
  - 신규 컬럼 추가
  - 기존 데이터 content_length 계산

- [ ] **키워드 추출 모듈**: `app/ml/keyword_extractor.py`
  - TF-IDF 기반 추출 로직
  - 한국어 불용어 처리
  - JSON 직렬화

- [ ] **통계 확인 스크립트**: `scripts/check_data_stats.py`
  - 전체 기사 수
  - 섹터별/도메인별 분포
  - 누락 필드 비율
  - 키워드 추출 완료율

### 1단계: 스키마 확장
- [x] `scripts/migrate_ml_columns.py` 작성
- [x] 마이그레이션 테스트 (테스트 DB)
- [x] 프로덕션 DB 마이그레이션 실행
- [x] 기존 데이터 content_length 계산

### 2단계: 키워드 추출 모듈
- [x] `app/ml/` 디렉토리 생성
- [x] `keyword_extractor.py` 작성
  - TF-IDF vectorizer 구현
  - 한국어 불용어 리스트 추가
  - 추출 로직 구현
- [x] 단위 테스트 작성
- [x] 기존 기사 키워드 일괄 추출

### 3단계: 수집 파이프라인 강화
- [x] RSS 파서 content 추출 로직 개선
- [x] 메타데이터 검증 로직 추가
- [x] 재시도 로직 구현
- [x] 통합 테스트

### 4단계: 통계 스크립트
- [x] `scripts/check_data_stats.py` 작성
- [x] 통계 출력 포맷 정의
- [x] 실행 및 결과 확인

## 📚 참고
- [MILESTONE4_PLAN.md](../MILESTONE4_PLAN.md): 전체 마일스톤 계획
- [scikit-learn TfidfVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html)

## ✅ 완료 조건
- [x] 스키마 마이그레이션 완료
- [x] 키워드 추출 모듈 동작 확인 (샘플 10개 테스트)
- [x] 통계 스크립트 실행 성공
- [ ] 1,000개 이상 기사 확보
- [ ] 메타데이터 완성도 90% 이상

---

**Labels**: `milestone-4`, `enhancement`, `day-1`
**Assignees**: @FerryLa
**Milestone**: 마일스톤 4 (ML 데이터 쌓기)
**Estimated Time**: 1 day

**Comment by La-vela** [comment]
이해 topic : 학습한 1000개 기사를 토대로 전체 IDF점수를 만들고 TF점수를 비교해서 중요도를 판단한다.
